### PR TITLE
Adds wrapper code for ClassList routines in Project

### PR DESCRIPTION
**Pull request by DrPaulSharp**

### CODE REVIEW

File: RAT/project.py [hunk 1]
```diff
@@ -170,5 +207,43 @@ def check_allowed_values(self, attribute: str, field_list: list[str], allowed_va
...
+        -------
+        wrapped_func : Callable
+            The wrapped routine.
+        """
+        @functools.wraps(func)
+        def wrapped_func(*args, **kwargs):
+            """Run the given function and then revalidate the "Project" model. If any exception is raised, restore
+            the previous state of the given ClassList and report details of the exception.
+            """
+            previous_state = copy.deepcopy(getattr(class_list, 'data'))
```
Comment (StephenNneji) [inline RAT/project.py:233]: Is a deep copy needed, I was thinking a shallow copy would work?

**Review by StephenNneji — APPROVED** [review]

File: RAT/project.py (hunk 1 above)
Comment (DrPaulSharp) [reply RAT/project.py:233]: No, a deep copy is required to copy the contents of the ClassList correctly.

Comment (StephenNneji) [reply RAT/project.py:233]: A shallow copy will have the order of the list but reference to the data, I thought the order of the list is what was needed. What happens when you shallow copy?

Comment (DrPaulSharp) [reply RAT/project.py:233]: If you shallow copy, you have a reference to the data. The code then changes the data, so if the data is changed in a way that proves invalid, the copy has a reference to the same changed data, and cannot restore the old version of the data as desired. The tests cover this case, simply changing "deepcopy" to "copy" will cause them to fail.

**Pull request merged**